simespresso/io_QE/espresso_class.py

```diff
@@ -0,0 +1,793 @@
+    def process_atomic_positions(self,f,pc,SP,units='(angstrom)'):
+        print('processing atomic_positions section')
+        try:
+            line = f.next()
+            SP[CUBA.KINEMATIC_VISCOSITY] = units
+            particle_list = []
+            while _ReadState.get_state(_ReadState.ATOMIC_POSITIONS,line) == _ReadState.ATOMIC_POSITIONS:
+                logging.debug('line in atomic positions section:'+str(line))
+                values = line.split()
+                print('values:'+str(values))
+                atom_pos = [0,0,0]
+                i = 0
+                if values[0] in atomtypes:
+                    atomtype = values[0]
+                    atom_id = i
+                    atom_pos[0] = float(values[1]) * 1e-10  #store position in meters; original in Angstrom
+                    atom_pos[1] = float(values[2]) * 1e-10
+                    atom_pos[2] = float(values[3]) * 1e-10
+                    s = str(i)
+                    #make uid using string of index
+                    #uidstring =
+                    #uid = uuid.UUID(uidstring)
+
+                    p = Particle([atom_pos[0],atom_pos[1],atom_pos[2]])  #,uuid.UUID(int=i)
+                    print('uid:'+str(p.uid))
+                    p.data[CUBA.CHEMICAL_SPECIE] = atomtype
+                    particle_list.append(p)
+                    try:
+                        line = f.next()
+                    except StopIteration:
+                        print('EOF')
+                        break
+            if len(particle_list):
+                pc.add_particles(particle_list)
+                  #  part_container.add_particle(p)
+                n=self.count_particles()
+                print('n_particles='+str(n))
+        except StopIteration:
+            return('EOF')
+
+        return pc
+
+    def count_particles(self):
+        n=0
+        for particle in self.pc.iter_particles():
+            n+=1
+        return n
+
+    def start_qe(self,name_in,name_out,path_to_espresso='/usr/bin/pw.x',mpi=False,mpi_Nprocessors=2):
+#        name_in = './test_pw.in'
+#        name_out = './test_pw.out'
+        if path_to_espresso is None:
+    #        path_to_espresso = '/usr/bin/pw.x'   #this appears to be the default install location for espresso
+            pass #I'll assume if no path given then pw.x is on path
+        if mpi:
+            command = 'mpirun -np '+str(mpi_Nprocessors)+' '+path_to_espresso+' < '+name_in +' > '+name_out
+        else:
+            command = path_to_espresso+' < '+name_in +' > '+name_out
+
+       # command = '/usr/bin/pw.x < '+name_in+' > '+name_out
+        print('qe wrapper attempting to run: '+command)
+        #alternative would be to use subprocess.check_call()  -  however this would give the same info as the
+        # try/except, while taking twice as long in the case of success, iiuc
+        try:
+            subprocess.Popen(command, shell=True, stdout=subprocess.PIPE).stdout.read()
```

Comment:
>> An exception should be raised if this doesn't work (i.e. can't find QE or something is with the input file..)

> I used a try/except instead of subprocess.check_call as the latter would take twice as long (in the case of success) while giving same error info in the case of failure

I don't understand why check_call would be slower, ?  have you experienced this?

Your updated code still does not raise an exception when things go wrong.  See https://travis-ci.org/simphony/simphony-espresso-dft/jobs/85990709#L647 where pw.x isn't found.  Here an exception should be thrown.

I don't think `Popen` throws an exception.   https://docs.python.org/2/library/subprocess.html#subprocess.Popen.  And if even if it did, you currently just catch the exception and write out the error.